src/app/utils/query-url-sanitization.ts

```diff
@@ -1,67 +1,99 @@
-const keyIdRegex = /\(\'\<?\{?[ ?0-9a-zA-Z-]*\}?\>?\'\)/g;
-
-// Matches patterns like "(query=<key>)" or "(itemat=<key>,mode=<mode>)"
-const functionParamInitialRegex = /(?<=\=).*?(?=,)/g;
-const functionParamFinalRegex = /(?<=\=).[^,]*(?=\))/g;
-
 // Matches pattterns within quotes e.g "displayName: Gupta"
 const quotedTextRegex = /"([^"]*)"/g;
 
-// Matches PII patterns
-const numberRegex =  /(?<=\/)\d+\b/g; // number between forward slashes
-const emailRegex = /([a-z0-9_\-.+]+)@\w+(\.\w+)*/gi;
-const guidRegex =
- /(\{){0,1}[0-9a-fA-F]{8}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{4}\-[0-9a-fA-F]{12}(\}){0,1}/gi;
-
-
 export function sanitizeQueryUrl(url: string): string {
   url = decodeURIComponent(url);
 
   // Extract query string
-  const { search } = parseSampleUrl(url);
+  const { search, queryVersion } = parseSampleUrl(url);
   const queryString: string = sanitizeQueryParameters(search);
 
   // Drop query string
   let sanitizedUrl = url.split('?')[0];
 
-  // Normalize parameters in param=<arbitraryKey> format with param={value} format
-  sanitizedUrl = sanitizedUrl.replace(functionParamInitialRegex, '<value>');
-  sanitizedUrl = sanitizedUrl.replace(functionParamFinalRegex, '<value>');
+  /*
+  * Remove url prefix and query version.
+  * They have special characters used in the check that follows
+  */
+  const urlPrefix = `${GRAPH_URL}/${queryVersion}`;
+  sanitizedUrl = sanitizedUrl.replace(urlPrefix, '');
 
-  // Replace IDs and ID placeholders with generic {id}
-  sanitizedUrl = sanitizedUrl.replace(keyIdRegex, '/<id>');
-  sanitizedUrl = sanitizedUrl.replace(guidRegex, '<guid>');
-  sanitizedUrl = sanitizedUrl.replace(numberRegex, '<number>');
+  const sections = sanitizedUrl.split('/');
+  sections.forEach(sec => {
+    if (containsIdentifier(sec)) {
+      const index = sections.indexOf(sec);
+      const replacementItemWithPrefix = `{${sections[index - 1]}-id}`;
+      sanitizedUrl = sanitizedUrl.replace(sec, replacementItemWithPrefix);
+    }
+  });
 
-  // Redact PII
-  sanitizedUrl = sanitizedUrl.replace(emailRegex, '<email>');
-  return `${sanitizedUrl}${queryString}` ;
+  return `${urlPrefix}${sanitizedUrl}${queryString}`;
 }
 
 function sanitizeQueryParameters(queryString: string): string {
-   const params = queryString.split('&');
-   let result: string = '';
-   if (params.length)
-   {
-      params.forEach(param => {
-       result += sanitizeQueryParameterValue(param) + '&';
-      });
-      result = result.slice(0, -1);
-   }
-   return result;
+  const params = queryString.split('&');
+  let result: string = '';
+  if (params.length) {
+    params.forEach(param => {
+      result += sanitizeQueryParameterValue(param) + '&';
+    });
+    result = result.slice(0, -1);
+  }
+  return result;
+}
+
+/*
+* A string contains an identifier if either:
+*   1. it is an alphanumeric string
+*   2. special characters are present
+*/
+export function containsIdentifier(segment: string): boolean {
```

Comment:
*Comment is for info purposes only, no action needed at this time besides opening a tracking work item to make key sanitization based on OData entity collections*

While I think it is safer to assume (but not totally safe) that a segment that has a special character is an identifier (I don't resource names will have a special character), I don't think it is safe to assume that an identifier has to have both alpha and numeric characters for an identifier. 

[From 4.1 Entity-Ids and Entity References](http://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part1-protocol.html#_Toc31358841): The entity-id MUST be an IRI as defined in [[RFC3987]](https://tools.ietf.org/html/rfc3987). It could be just a number, it could be just a letter. There is a Python package for validating this form: https://pypi.org/project/rfc3987/. I didn't find one on npm.

Reply:
Noted. I have recorded this in #745